Mathlib/GroupTheory/GroupAction/Jordan.lean

```diff
@@ -0,0 +1,673 @@
+      rw [← Nat.succ_lt_succ_iff, ← hsn, Nat.succ_pred this]
+      refine ⟨?_, rfl⟩
+      apply Set.ncard_lt_ncard _ (Set.toFinite s)
+      exact ⟨Set.inter_subset_left, fun h ↦ hgb (Set.inter_subset_right (h hb))⟩
+    apply Set.ncard_ne_zero_of_mem (a := a)
+    exact ⟨ha, hga⟩
+
+  · -- CASE : 2 * s.ncard ≥ Nat.card α
+    have : Set.Nontrivial sᶜ := by
+      rw [← Set.one_lt_encard_iff_nontrivial, ← sᶜ.toFinite.cast_ncard_eq, Nat.one_lt_cast]
+      rw [← Nat.add_lt_add_iff_left, Set.ncard_add_ncard_compl, add_comm, hsn]
+      exact hsn'
+    -- get a, b ∈ sᶜ, a ≠ b
+    obtain ⟨a, ha : a ∈ sᶜ, b, hb : b ∈ sᶜ, hab⟩ := this
+
+    obtain ⟨g, hga, hgb⟩ := exists_mem_smul_and_notMem_smul (G := G)
+      sᶜ.toFinite (Set.nonempty_of_mem ha)
+      (fun h ↦ by simp only [Set.compl_univ_iff.mp h, Set.not_nonempty_empty] at hs_nonempty)
+      hab
+    let t := s ∩ g • s
+    have : a ∉ s ∪ g • s := by
+      rw [Set.mem_union]
+      intro h
+      rcases h with h | h
+      · exact ha h
+      · rw [Set.smul_set_compl] at hga; exact hga h
+    have ht_trans : IsPretransitive (fixingSubgroup G t) (ofFixingSubgroup G t) :=
+        IsPretransitive.isPretransitive_ofFixingSubgroup_inter hs_trans
+          (fun h ↦ this (by rw [h]; trivial))
+
+    suffices ∃ m : ℕ, m < n ∧ t.ncard = Nat.succ m by
+      obtain ⟨m, hmn, htm⟩ := this
+      exact hrec m hmn hG htm
+        (lt_trans (by rwa [add_lt_add_iff_left, Nat.succ_lt_succ_iff]) hsn') ht_trans
+
+    -- from : t ⊆ s, a ∈ t, b ∉ t,
+    -- have : 1 ≤ t.ncard < fintype.card s
+    use t.ncard.pred
+    suffices  t.ncard ≠ 0 by
+      rw [← Nat.succ_lt_succ_iff, ← hsn, Nat.succ_pred this]
+      refine ⟨?_, rfl⟩
+      apply Set.ncard_lt_ncard _ (Set.toFinite s)
+      rw [Set.ssubset_def]
+      refine ⟨Set.inter_subset_left, fun h ↦ hb ?_⟩
+      suffices s = g • s by
+        rw [this]
+        simpa only [Set.smul_set_compl, Set.mem_compl_iff, Set.not_notMem] using hgb
+      apply Set.eq_of_subset_of_ncard_le _ _ (g • s).toFinite
+      · exact subset_trans h Set.inter_subset_right
+      · rw [Set.ncard_smul_set]
+    · rw [← Nat.pos_iff_ne_zero, Set.ncard_pos]
+      -- Pigeon hole lemma
+      apply Set.ncard_pigeonhole'
+      · rw [Set.ncard_smul_set, ← two_mul, hsn]; exact hn2
+      · exact fun h ↦ this (by rw [h]; trivial)
+
+/- -- TODO : prove
+theorem strong_jordan_of_preprimitive (hG : IsPreprimitive G α)
+    {s : Set α} {n : ℕ} (hsn : s.ncard = n.succ)
+    (hsn' : 1 + n.succ < Nat.card α)
+    (hs_prim : IsPreprimitive (fixingSubgroup G s) (ofFixingSubgroup G s)) :
+    IsMultiplyPreprimitive (Subgroup.normalClosure (fixingSubgroup G s : Set G)) α 2 :=
+  sorry
+ -/
+
+theorem is_two_preprimitive_weak_jordan [DecidableEq α]
+    (hG : IsPreprimitive G α) {s : Set α} {n : ℕ}
+    (hsn : s.ncard = n.succ) (hsn' : 1 + n.succ < Nat.card α)
+    (hs_prim : IsPreprimitive (fixingSubgroup G s) (ofFixingSubgroup G s)) :
+    IsMultiplyPreprimitive G α 2 := by
```

Comment:
This proof is quite similar to the proof above. Is it not possible to combine the proofs into one?